### DISCUSSION TITLE
Custom Game

### DISCUSSION THREAD

**Discussion by VictorXLR** [trackmania-rl/tmrl#42] [Q&A]
Hi Yann,

I really like your library, I have been trying to port it to play a much quicker game, currently OLDTV 
https://store.steampowered.com/app/643270/OLDTV/ and I would like your feedback on how to setup the reward function and action_states. 

I looked at the guide you provided for inputs and reward function but i am unable to apply it for something like OLDTV. I am very new to this field so any advice would be appreciated

the code i have been following is mainly this: https://github.com/trackmania-rl/tmrl/blob/master/tmrl/custom/custom_gym_interfaces.py

thank you

Victor

**Comment by yannbouteiller** [comment]
Hi Victor, I have converted this post into a discussion since this is not an issue with `tmrl` :)

When designing an RL environment in a video game, the reward function is the trickiest part.

The action_space should be fairly straightforward, as this is basically the possible controls that your policy can output. For instance, in the `TrackMania` environment, actions are the pressure on the gas pedal, the pressure on the brake pedal, and the steering angle. We normalize all those outputs between -1.0 and +1.0, thus the action_space is `gymnasium.spaces.Box(low=-1.0, high=1.0, shape=(3, ))`, representing a numpy array of 3 values between -1.0 and +1.0. You can do something similar with any kind of actions you would like your policy to output.

The reward function, however, is tricky for several reasons:
- You need to measure how well your agent is performing, which is often tricky in itself in a video game,
- The sparser your reward signal, the more difficult it is for conventional RL algorithms to optimize your policy; for instance your could in theory give a reward of 0 every timestep and a reward of 1 on the only timestep when your agent completes the task, but since RL algorithms are typically based on random exploration it may be extremely difficult for the agent to even find the state where it completes the task. In `TrackMania`, we use the OpenPlanet API to retrieve the 3D position of the car, which enables us to compute a meaningful reward at every timestep.

Concerning OLDTV, I never played the game, but as far as I understand it would not make much sense to apply a Reinforcement Learning approach to this in my opinion, because your actions do not influence the environment and the reward is immediate: either you click as you are supposed to and you win, either you misclick and lose. This is typically a supervised learning problem: you can collect a dataset by playing, and train an algorithm to classify screenshots as "right click" or "left click". The main difficulty will be for your model to process screenshots in order to extract text meaning and colors. This is a super interesting problem in terms of computer vision though!

That being said, it may become a super interesting RL problem and a very good fit for `tmrl` if you want your agent to play in real-time, as in clicking as fast as possible but only when it knows the answer (because at the beginning of each round we don't see the text yet, we have to wait for it to move toward the center of the screen). In this situation, you would have 3 actions: "wait", "right click" and "left click". An episode would start at each round, and end when you decide to take one of the clicking action. A success would give you a reward of +1,0, and a failure would give you a reward of -1.0, whereas waiting would give you a reward of 0.0 (or a small penalty but careful with that: you don't want your agent to fall in the situation where it clicks randomly as fast as possible in order to keep the penalty from building up). Typically, your policy would be a neural network with 3 outputs that takes screenshots as input, and you would select the output with the highest value at each time step.

So, essentially, what you need is a way of telling whether you failed or passed after clicking. Note that providing a reward only at the timestep where you click makes sense in terms of completing the task as fast as possible because of the "discount factor" that we call "gamma" in RL. I can explain if you don't see why.

**Reply by VictorXLR** [reply]
Hi, 

Thanks so much, its a really unique game due to the fact it needs to solve both text recognition and color recognition in the same time and i tried using openCV to solve it but couldnt approach cases where the background would change or the text would be rotated

Would you have any supporting starting points for someone doing something like this..

Currently using the win32 api to find the window and sending clicks to the application but ill love more training material regarding applying supervised learning to it.

Thank you

Victor

**Reply by yannbouteiller** [reply]
So, the field of vision/text dual embeddings is very hot right now, in particular due to a model named CLIP, and extremely sophisticated approaches exist to tackle your problem, based on visual transformers. For instance you have models such as [this](https://huggingface.co/docs/transformers/model_doc/trocr) and [this](https://huggingface.co/docs/transformers/v4.27.1/en/model_doc/vision-text-dual-encoder#transformers.VisionTextDualEncoderModel) that seem fit as a starting point if you had a lot of experience in the field.

However, these approaches are very difficult to understand and implement, and honestly you should not try this for now if you are a beginner. Instead, I highly suggest you start with implementing a simple Convolutional Neural Network to classify your frames, as CNNs are much easier to understand and serve as a building block for many of these approaches. You should find easy tutorials classifying the MNIST digits for this, which is very relevant to your setting. Maybe you can then quickly move on to the more advanced CNNs for image classification such as ResNets, and keep experimenting with more and more complex models.

The very cool point in your setting is that it is toyish enough to experiment with, and yet has this dual text/image recognition at its heart, which is super relevant to human vision.

**Reply by VictorXLR** [reply]
Thank you so much for everything. I will start working on a CNN and other implementations before coming back to this project. 

I appreciate all the suggestions and I will try to be much more active as i try and solve this project.

Have a great week ahead

**Reply by VictorXLR** [reply]
Also I took a machine learning class in university working on this dataset and its all coming back to me, will keep send updates of my progress and reach out if i have any issues

**Comment by yannbouteiller** [comment]
Also I forgot to mention, but since your action space is discrete (only 3 actions from which you need to pick 1 at each time step), you may like to experiment with DQN before moving to more advanced algorithms, as you said you are new to the field :)
